File: packages/agent-governance-dotnet/src/AgentGovernance/Mcp/McpGateway.cs
// Copyright (c) Microsoft Corporation. Licensed under the MIT License.

using System.Diagnostics;
using System.Text.Json;
using System.Text.RegularExpressions;
using AgentGovernance.Audit;
using AgentGovernance.Integration;
using AgentGovernance.Mcp.Abstractions;
using AgentGovernance.Policy;
using AgentGovernance.RateLimiting;
using AgentGovernance.Telemetry;
using Microsoft.Extensions.Logging;

namespace AgentGovernance.Mcp;

/// <summary>
/// MCP governance gateway that intercepts tool calls through a 5-stage pipeline:
/// <list type="number">
///   <item><b>Deny-list</b> — Immediately block tools on the deny list.</item>
///   <item><b>Allow-list</b> — If an allow-list is configured, only permit listed tools.</item>
///   <item><b>Parameter sanitization</b> — Scan parameters for dangerous patterns (PII, shell injection).</item>
///   <item><b>Rate limiting</b> — Enforce per-agent call budgets.</item>
///   <item><b>Human approval</b> — Route sensitive tool calls through human-in-the-loop review.</item>
/// </list>
/// <para>
/// The gateway is <b>fail-closed</b>: any exception during pipeline evaluation results in denial.
/// Integrates with the existing <see cref="GovernanceKernel"/> policy engine and rate limiter.
/// </para>
/// </summary>
/// <remarks>
/// Ported from the Python <c>MCPGateway</c> in <c>agent_os/mcp_gateway.py</c>.
/// </remarks>
public sealed class McpGateway
{
    private readonly GovernanceKernel _kernel;
    private readonly HashSet<string> _deniedTools;
    private readonly HashSet<string> _allowedTools;
    private readonly HashSet<string> _sensitiveTools;
    private readonly bool _enableBuiltinSanitization;
    private readonly Func<string, string, Dictionary<string, object>, ApprovalStatus>? _approvalCallback;
    private readonly bool _requireHumanApproval;
    private readonly bool _enableCredentialRedaction;
    private readonly IMcpAuditSink _auditSink;
    private readonly TimeProvider _timeProvider;

    /// <summary>
    /// Maximum tool calls per agent before rate-limiting kicks in.
    /// Set to <c>0</c> or negative to disable budget-based rate limiting
    /// (the kernel's policy-based rate limiter still applies).
    /// When a <see cref="RateLimiter"/> is configured, this value is informational only —
    /// the limiter's <see cref="McpSlidingRateLimiter.MaxCallsPerWindow"/> controls the actual limit.
    /// </summary>
    public int MaxToolCallsPerAgent { get; init; } = 1000;

    /// <summary>
    /// Optional sliding-window rate limiter. When set, replaces the simple counter-based
    /// budget with a proper sliding window that automatically expires old calls.
    /// </summary>
    public McpSlidingRateLimiter? RateLimiter { get; set; }

    /// <summary>
    /// Optional <see cref="GovernanceMetrics"/> instance for recording
    /// telemetry from the MCP gateway pipeline.
    /// </summary>
    public GovernanceMetrics? Metrics { get; set; }

    /// <summary>
    /// Optional logger for recording gateway decisions and errors.
    /// When <c>null</c>, no logging occurs — the gateway operates silently.
    /// </summary>
    public ILogger<McpGateway>? Logger { get; set; }

    /// <summary>
    /// Initializes a new <see cref="McpGateway"/>.
    /// </summary>
    /// <param name="kernel">
    /// The <see cref="GovernanceKernel"/> whose policy engine and rate limiter will be used.
    /// </param>
    /// <param name="deniedTools">Tools that are always blocked, regardless of policy.</param>
    /// <param name="allowedTools">
    /// If non-empty, only these tools are permitted (allow-list mode).
    /// An empty or <c>null</c> list disables the allow-list filter.
    /// </param>
    /// <param name="sensitiveTools">Tools that require human approval even if policy allows them.</param>
    /// <param name="approvalCallback">
    /// Optional callback for human-in-the-loop approval.
    /// Signature: <c>(agentId, toolName, parameters) → ApprovalStatus</c>.
    /// </param>
    /// <param name="enableBuiltinSanitization">
    /// Whether to apply built-in dangerous-pattern sanitization (SSN, credit cards, shell injection).
    /// Defaults to <c>true</c>.
    /// </param>
    /// <param name="requireHumanApproval">
    /// When <c>true</c>, ALL tool calls require human approval (not just sensitive tools).
    /// Defaults to <c>false</c>.
    /// </param>
    /// <param name="enableCredentialRedaction">
    /// Whether to redact credentials before audit entries are stored.
    /// Defaults to <c>true</c>.
    /// </param>
    /// <param name="auditSink">The sink used to persist audit entries.</param>
    /// <param name="timeProvider">The clock used for audit timestamps.</param>
    public McpGateway(
        GovernanceKernel kernel,
        IEnumerable<string>? deniedTools = null,
        IEnumerable<string>? allowedTools = null,
        IEnumerable<string>? sensitiveTools = null,
        Func<string, string, Dictionary<string, object>, ApprovalStatus>? approvalCallback = null,
        bool enableBuiltinSanitization = true,
        bool requireHumanApproval = false,
        bool enableCredentialRedaction = true,
        IMcpAuditSink? auditSink = null,
        TimeProvider? timeProvider = null)
    {
        ArgumentNullException.ThrowIfNull(kernel);

        _kernel = kernel;
        _deniedTools = deniedTools is not null
            ? new HashSet<string>(deniedTools, StringComparer.OrdinalIgnoreCase)
            : new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        _allowedTools = allowedTools is not null
            ? new HashSet<string>(allowedTools, StringComparer.OrdinalIgnoreCase)
            : new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        _sensitiveTools = sensitiveTools is not null
            ? new HashSet<string>(sensitiveTools, StringComparer.OrdinalIgnoreCase)
            : new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        _approvalCallback = approvalCallback;
        _enableBuiltinSanitization = enableBuiltinSanitization;
        _requireHumanApproval = requireHumanApproval;
        _enableCredentialRedaction = enableCredentialRedaction;
        _auditSink = auditSink ?? new InMemoryMcpAuditSink();
        _timeProvider = timeProvider ?? TimeProvider.System;
    }

    /// <summary>
    /// Intercepts an MCP tool call and runs it through the 5-stage governance pipeline.
    /// </summary>
    /// <param name="agentId">The agent's DID.</param>
    /// <param name="toolName">Name of the MCP tool being called.</param>
    /// <param name="parameters">Parameters being passed to the tool.</param>
    /// <returns>
    /// A tuple of (allowed, reason). If <c>allowed</c> is <c>false</c>,
    /// the tool call should be blocked.
    /// </returns>
    public (bool Allowed, string Reason) InterceptToolCall(
        string agentId,
        string toolName,
        Dictionary<string, object> parameters)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(agentId);
        ArgumentException.ThrowIfNullOrWhiteSpace(toolName);
        parameters ??= new Dictionary<string, object>();

        var sw = Stopwatch.StartNew();
        Logger?.LogInformation("MCP tool call intercepted: {ToolName} by {AgentId}", toolName, agentId);

        try
        {
            var (allowed, reason, approvalStatus) = Evaluate(agentId, toolName, parameters);

            sw.Stop();
            var stage = DetermineStage(allowed, reason);
            var rateLimited = reason.Contains("exceeded call budget", StringComparison.OrdinalIgnoreCase)
                           || reason.Contains("rate limit", StringComparison.OrdinalIgnoreCase);
            Metrics?.RecordMcpDecision(allowed, agentId, toolName, sw.Elapsed.TotalMilliseconds, stage, rateLimited);

            if (allowed)
            {
                Logger?.LogInformation("MCP tool call allowed: {ToolName} for {AgentId}", toolName, agentId);
            }
            else
            {
                Logger?.LogWarning("MCP tool call denied: {ToolName} for {AgentId} - {Reason}", toolName, agentId, reason);
            }

            RecordAuditEntry(agentId, toolName, parameters, allowed, reason, approvalStatus);

            return (allowed, reason);
        }
        catch (Exception ex)
        {
            sw.Stop();
            Logger?.LogError(ex, "MCP gateway error for {ToolName} - failing closed", toolName);

            // Fail-closed: any exception → deny.
            var failReason = "Gateway error (fail-closed).";

            Metrics?.RecordMcpDecision(false, agentId, toolName, sw.Elapsed.TotalMilliseconds, "error");

            try
            {
                RecordAuditEntry(agentId, toolName, parameters, false, failReason, null);
            }
            catch (Exception auditEx)
            {
                Logger?.LogError(auditEx, "MCP audit sink failure while recording a fail-closed decision");
            }

            return (false, failReason);
        }
    }

    /// <summary>
    /// Returns a defensive copy of the audit log.
    /// </summary>
    public IReadOnlyList<McpAuditEntry> AuditLog
    {
        get
        {
            return _auditSink is InMemoryMcpAuditSink inMemoryAuditSink
                ? inMemoryAuditSink.GetSnapshot()
                : Array.Empty<McpAuditEntry>();
        }
    }

    /// <summary>
    /// Returns the current call count for an agent.
    /// When a sliding window <see cref="RateLimiter"/> is configured,
    /// returns the count of calls within the current window.
    /// </summary>
    public int GetAgentCallCount(string agentId)
    {
        if (RateLimiter is not null)
        {
            return RateLimiter.GetCallCount(agentId);
        }

        return 0;
    }

    /// <summary>
    /// Resets the call budget for a specific agent.
    /// </summary>
    public void ResetAgentBudget(string agentId)
    {
        if (RateLimiter is not null)
        {
            RateLimiter.Reset(agentId);
        }
    }

    /// <summary>
    /// Resets call budgets for all agents.
    /// </summary>
    public void ResetAllBudgets()
    {
        if (RateLimiter is not null)
        {
            RateLimiter.ResetAll();
        }
    }

    // ── 5-Stage Pipeline ─────────────────────────────────────────────────

    private (bool Allowed, string Reason, ApprovalStatus? Status) Evaluate(
        string agentId,
        string toolName,
        Dictionary<string, object> parameters)
    {
        // Stage 1: Deny-list check
        if (_deniedTools.Contains(toolName))
        {
            return (false, $"Tool '{toolName}' is on the deny list", null);
        }

        // Stage 2: Allow-list check (empty allow-list = all tools allowed)
        if (_allowedTools.Count > 0 && !_allowedTools.Contains(toolName))
        {
            return (false, $"Tool '{toolName}' is not on the allow list", null);
        }

        // Stage 3: Parameter sanitization
        var sanitizationResult = SanitizeParameters(parameters);
        if (!sanitizationResult.Clean)
        {
            return (false, $"Parameters matched dangerous pattern: {sanitizationResult.MatchedPattern}", null);
        }

        // Also evaluate through the kernel's policy engine for policy-based blocking.
        var policyResult = _kernel.EvaluateToolCall(agentId, toolName, parameters);
        if (!policyResult.Allowed)
        {
            return (false, policyResult.Reason, null);
        }

        // Stage 4: Rate limiting (sliding window or disabled)
        if (RateLimiter is not null)
        {
            // Peek — don't consume a permit yet (we may need human approval first).
            var remaining = RateLimiter.GetRemainingBudget(agentId);
            if (remaining <= 0)
            {
                return (false, $"Agent '{agentId}' exceeded call budget ({RateLimiter.MaxCallsPerWindow}/{RateLimiter.MaxCallsPerWindow})", null);
            }
        }

        // Stage 5: Human approval
        if (_requireHumanApproval || _sensitiveTools.Contains(toolName))
        {
            var approvalResult = EvaluateHumanApproval(agentId, toolName, parameters);
            // Only consume a rate-limit permit on approved calls
            if (approvalResult.Allowed && RateLimiter is not null)
            {
                if (!RateLimiter.TryAcquire(agentId))
                {
                    // Race: another thread consumed the last permit between check and acquire.
                    return (false, $"Agent '{agentId}' exceeded call budget ({RateLimiter.MaxCallsPerWindow}/{RateLimiter.MaxCallsPerWindow})", null);
                }
            }
            return approvalResult;
        }

        // Consume a rate-limit permit for calls that are allowed without human approval
        if (RateLimiter is not null)
        {
            if (!RateLimiter.TryAcquire(agentId))
            {
                return (false, $"Agent '{agentId}' exceeded call budget ({RateLimiter.MaxCallsPerWindow}/{RateLimiter.MaxCallsPerWindow})", null);
            }
        }

        return (true, "Allowed by policy", null);
    }

    private (bool Allowed, string Reason, ApprovalStatus? Status) EvaluateHumanApproval(
        string agentId,
        string toolName,
        Dictionary<string, object> parameters)
    {
        if (_approvalCallback is null)
        {
            return (false, "Awaiting human approval", ApprovalStatus.Pending);
        }

        try
        {
            var status = _approvalCallback(agentId, toolName, parameters);

            return status switch
            {
                ApprovalStatus.Approved => (true, "Approved by human reviewer", ApprovalStatus.Approved),
                ApprovalStatus.Denied => (false, "Human approval denied", ApprovalStatus.Denied),
                ApprovalStatus.Pending => (false, "Awaiting human approval", ApprovalStatus.Pending),
                _ => (false, "Unknown approval status — fail-closed", null)
            };
        }
        catch
        {
            // Fail-closed: approval callback error → deny.
            return (false, "Approval callback error — fail-closed", ApprovalStatus.Denied);
        }
    }

    private static string DetermineStage(bool allowed, string reason)
    {
        if (allowed)
            return "allowed";
        if (reason.Contains("deny list", StringComparison.OrdinalIgnoreCase))
            return "deny_list";
        if (reason.Contains("allow list", StringComparison.OrdinalIgnoreCase))
            return "allow_list";
        if (reason.Contains("dangerous pattern", StringComparison.OrdinalIgnoreCase)
            || reason.Contains("sanitiz", StringComparison.OrdinalIgnoreCase))
            return "sanitization";
        if (reason.Contains("exceeded call budget", StringComparison.OrdinalIgnoreCase)
            || reason.Contains("rate limit", StringComparison.OrdinalIgnoreCase))
            return "rate_limit";
        if (reason.Contains("approval", StringComparison.OrdinalIgnoreCase))
            return "approval";
        return "policy";
    }

    private static (bool Clean, string? MatchedPattern) SanitizeParameters(Dictionary<string, object> parameters)
    {
        if (parameters.Count == 0)
            return (true, null);

        string paramText;
        try
        {
            paramText = JsonSerializer.Serialize(parameters);
        }
        catch
        {
            paramText = string.Join(" ", parameters.Values.Select(v => v?.ToString() ?? string.Empty));
        }

        foreach (var (pattern, name) in SanitizationDefaults.AllPatterns)
        {
            try
            {
                if (pattern.IsMatch(paramText))
                {
                    return (false, name);
                }
            }
            catch (RegexMatchTimeoutException)
            {
                // Fail-closed: regex timeout → deny.
                return (false, $"{name} (regex timeout)");
            }
        }

        return (true, null);
    }

    private void RecordAuditEntry(
        string agentId,
        string toolName,
        Dictionary<string, object> parameters,
        bool allowed,
        string reason,
        ApprovalStatus? approvalStatus)
    {
        var auditParameters = _enableCredentialRedaction
            ? CredentialRedactor.RedactDictionary(parameters)
            : new Dictionary<string, object>(parameters);

        _auditSink.RecordAsync(new McpAuditEntry
        {
            Timestamp = _timeProvider.GetUtcNow(),
            AgentId = agentId,
            ToolName = toolName,
            Parameters = auditParameters,
            Allowed = allowed,
            Reason = reason,
            ApprovalStatus = approvalStatus
        }).GetAwaiter().GetResult();
    }
}

/// <summary>
/// A single audit entry recorded by the <see cref="McpGateway"/>.
/// </summary>
public sealed class McpAuditEntry
{
    /// <summary>When the evaluation occurred.</summary>
    public DateTimeOffset Timestamp { get; init; }

    /// <summary>The agent's DID.</summary>
    public required string AgentId { get; init; }

    /// <summary>The tool that was called.</summary>
    public required string ToolName { get; init; }

    /// <summary>Parameters passed to the tool.</summary>
    public Dictionary<string, object> Parameters { get; init; } = new();

    /// <summary>Whether the call was allowed.</summary>
    public bool Allowed { get; init; }

    /// <summary>Reason for the decision.</summary>
    public required string Reason { get; init; }

    /// <summary>Human approval status, if applicable.</summary>
    public ApprovalStatus? ApprovalStatus { get; init; }
}

File: packages/agent-mesh/sdks/typescript/src/mcp-gateway.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { CredentialRedactor } from './credential-redactor';
import { MCPSlidingRateLimiter } from './mcp-sliding-rate-limiter';
import {
  ApprovalStatus,
  MCPAuditSink,
  MCPGatewayAuditEntry,
  MCPGatewayConfig,
  MCPGatewayDecisionResult,
  MCPMaybePromise,
  MCPResponseFinding,
  MCPSlidingRateLimitResult,
  MCPWrappedServerConfig,
} from './types';
import {
  createRegexScanBudget,
  debugSecurityFailure,
  DEFAULT_MCP_CLOCK,
  validateRegex,
} from './mcp-utils';

const BUILTIN_DANGEROUS_PATTERNS = [
  /\b\d{3}-\d{2}-\d{4}\b/gi,
  /\b\d{4}[- ]?\d{4}[- ]?\d{4}[- ]?\d{4}\b/gi,
  /;\s*(?:rm|del|format|mkfs)\b/gi,
  /`[^`]+`/g,
];

export class InMemoryMCPAuditSink implements MCPAuditSink {
  private readonly entries: MCPGatewayAuditEntry[] = [];

  record(entry: MCPGatewayAuditEntry): void {
    this.entries.push(entry);
  }

  getEntries(): MCPGatewayAuditEntry[] {
    return [...this.entries];
  }
}

export class MCPGateway {
  private readonly deniedTools: string[];
  private readonly allowedTools: string[];
  private readonly sensitiveTools: string[];
  private readonly blockedPatterns: Array<string | RegExp>;
  private readonly enableBuiltinSanitization: boolean;
  private readonly approvalHandler?: MCPGatewayConfig['approvalHandler'];
  private readonly policyEvaluator?: MCPGatewayConfig['policyEvaluator'];
  private readonly metrics?: MCPGatewayConfig['metrics'];
  private readonly auditSink?: MCPAuditSink;
  private readonly clock: NonNullable<MCPGatewayConfig['clock']>;
  private readonly logger: MCPGatewayConfig['logger'];
  private readonly scanTimeoutMs: number;
  private readonly rateLimiter?: {
    consume(agentId: string): MCPMaybePromise<MCPSlidingRateLimitResult>;
  };
  private readonly redactor = new CredentialRedactor();
  private readonly auditEntries: MCPGatewayAuditEntry[] = [];

  constructor(config: MCPGatewayConfig = {}) {
    this.deniedTools = config.deniedTools ?? [];
    this.allowedTools = config.allowedTools ?? [];
    this.sensitiveTools = config.sensitiveTools ?? [];
    this.blockedPatterns = config.blockedPatterns ?? [];
    this.enableBuiltinSanitization = config.enableBuiltinSanitization ?? true;
    this.approvalHandler = config.approvalHandler;
    this.policyEvaluator = config.policyEvaluator;
    this.metrics = config.metrics;
    this.auditSink = config.auditSink;
    this.clock = config.clock ?? DEFAULT_MCP_CLOCK;
    this.logger = config.logger;
    this.scanTimeoutMs = config.scanTimeoutMs ?? 100;
    this.blockedPatterns.forEach((pattern) => {
      if (pattern instanceof RegExp) {
        validateRegex(pattern);
      }
    });
    this.rateLimiter =
      config.rateLimiter
      ?? (config.rateLimit
        ? new MCPSlidingRateLimiter(config.rateLimit)
        : undefined);
  }

  async evaluateToolCall(
    agentId: string,
    toolName: string,
    params: Record<string, unknown> = {},
  ): Promise<MCPGatewayDecisionResult> {
    try {
      return await this.evaluate(agentId, toolName, params);
    } catch (error) {
      debugSecurityFailure(this.logger, 'gateway.evaluateToolCall', error);
      return this.finalize(
        agentId,
        toolName,
        params,
        {
          allowed: false,
          reason: 'Internal gateway error - access denied (fail closed)',
          auditParams: this.redactor.redact(params).redacted as Record<string, unknown>,
          findings: [],
        },
      );
    }
  }

  get auditLog(): MCPGatewayAuditEntry[] {
    return [...this.auditEntries];
  }

  static wrapServer(
    serverConfig: Record<string, unknown>,
    config: MCPGatewayConfig = {},
  ): MCPWrappedServerConfig {
    return {
      serverConfig: { ...serverConfig },
      allowedTools: [...(config.allowedTools ?? [])],
      deniedTools: [...(config.deniedTools ?? [])],
      sensitiveTools: [...(config.sensitiveTools ?? [])],
      rateLimit: config.rateLimit,
    };
  }

  private async evaluate(
    agentId: string,
    toolName: string,
    params: Record<string, unknown>,
  ): Promise<MCPGatewayDecisionResult> {
    const auditParams = this.redactor.redact(params).redacted as Record<string, unknown>;
    const findings: MCPResponseFinding[] = [];

    if (this.deniedTools.includes(toolName)) {
      return this.finalize(agentId, toolName, params, {
        allowed: false,
        reason: `Tool '${toolName}' is on the deny list`,
        auditParams,
        findings,
      });
    }

    if (
      this.allowedTools.length > 0
      && !this.allowedTools.includes(toolName)
    ) {
      return this.finalize(agentId, toolName, params, {
        allowed: false,
        reason: `Tool '${toolName}' is not on the allow list`,
        auditParams,
        findings,
      });
    }

    const policyDecision = this.policyEvaluator?.evaluate(toolName, {
      agentId,
      ...params,
    });
    if (policyDecision === 'deny') {
      return this.finalize(agentId, toolName, params, {
        allowed: false,
        reason: `Policy denied tool '${toolName}'`,
        auditParams,
        findings,
        policyDecision,
      });
    }

    const sanitizationFinding = this.checkSanitization(params);
    if (sanitizationFinding) {
      findings.push(sanitizationFinding);
      return this.finalize(agentId, toolName, params, {
        allowed: false,
        reason: sanitizationFinding.message,
        auditParams,
        findings,
        policyDecision,
      });
    }

    const rateLimit = await this.rateLimiter?.consume(agentId);
    if (rateLimit && !rateLimit.allowed) {
      this.metrics?.recordMcpRateLimitHit(agentId, {
        toolName,
        retryAfterMs: rateLimit.retryAfterMs,
      });
      return this.finalize(agentId, toolName, params, {
        allowed: false,
        reason: `Agent '${agentId}' exceeded the MCP rate limit`,
        auditParams,
        findings,
        policyDecision,
        rateLimit,
      });
    }

    const requiresApproval =
      this.sensitiveTools.includes(toolName)
      || policyDecision === 'review';
    if (requiresApproval) {
      let approvalStatus = ApprovalStatus.Pending;
      if (this.approvalHandler) {
        approvalStatus = (await this.approvalHandler({
          agentId,
          toolName,
          params,
          auditParams,
          findings,
          policyDecision,
        })) ?? ApprovalStatus.Pending;
      }

      if (approvalStatus !== ApprovalStatus.Approved) {
        const reason = approvalStatus === ApprovalStatus.Denied
          ? 'Human approval denied'
          : 'Awaiting human approval';
        return this.finalize(agentId, toolName, params, {
          allowed: false,
          reason,
          auditParams,
          findings,
          approvalStatus,
          policyDecision,
          rateLimit,
        });
      }

      return this.finalize(agentId, toolName, params, {
        allowed: true,
        reason: 'Approved by human reviewer',
        auditParams,
        findings,
        approvalStatus,
        policyDecision,
        rateLimit,
      });
    }

    return this.finalize(agentId, toolName, params, {
      allowed: true,
      reason: 'Allowed by policy',
      auditParams,
      findings,
      policyDecision,
      rateLimit,
    });
  }

  private finalize(
    agentId: string,
    toolName: string,
    params: Record<string, unknown>,
    decision: MCPGatewayDecisionResult,
  ): MCPGatewayDecisionResult {
    const entry = {
      timestamp: new Date().toISOString(),
      agentId,
      toolName,
      params: decision.auditParams,
      auditParams: decision.auditParams,
      allowed: decision.allowed,
      reason: decision.reason,
      approvalStatus: decision.approvalStatus,
      policyDecision: decision.policyDecision,
      findings: decision.findings,
    };

    this.auditEntries.push(entry);
    this.auditSink?.record(entry);

    this.metrics?.recordMcpDecision(decision.allowed ? 'allow' : 'deny', {
      agentId,
      toolName,
      reason: decision.reason,
      approvalStatus: decision.approvalStatus ?? '',
    });

    if (decision.findings.length > 0) {
      this.metrics?.recordMcpThreatsDetected(decision.findings.length, {
        agentId,
        toolName,
      });
    }

    return decision;
  }

  private checkSanitization(
    params: Record<string, unknown>,
  ): MCPResponseFinding | undefined {
    const serialized = JSON.stringify(params);
    const budget = createRegexScanBudget(this.clock, this.scanTimeoutMs);

    for (const pattern of this.blockedPatterns) {
      budget.checkpoint('Regex scan exceeded time budget - access denied');
      if (
        (typeof pattern === 'string' && serialized.includes(pattern))
        || (pattern instanceof RegExp && pattern.test(serialized))
      ) {
        return {
          type: 'imperative_language',
          severity: 'critical',
          message: `Parameters matched blocked pattern: ${String(pattern)}`,
          matchedText: String(pattern),
          path: '$',
        };
      }
    }

    if (this.enableBuiltinSanitization) {
      if (hasShellExpansion(serialized)) {
        return {
          type: 'imperative_language',
          severity: 'critical',
          message: 'Parameters matched dangerous pattern: shell_expansion',
          matchedText: 'shell_expansion',
          path: '$',
        };
      }

      for (const pattern of BUILTIN_DANGEROUS_PATTERNS) {
        budget.checkpoint('Regex scan exceeded time budget - access denied');
        if (pattern.test(serialized)) {
          return {
            type: 'imperative_language',
            severity: 'critical',
            message: `Parameters matched dangerous pattern: ${pattern.source}`,
            matchedText: pattern.source,
            path: '$',
          };
        }
      }
    }

    return undefined;
  }
}

function hasShellExpansion(value: string): boolean {
  const startIndex = value.indexOf('$(');
  return startIndex !== -1 && value.indexOf(')', startIndex + 2) !== -1;
}
